### PR TITLE
Mypy type issues part 2

### PR DESCRIPTION
**Pull request by dstrain115**
- Continue fixing mypy type issues.

Some highlights include:
- Fix bloq_example decorator to pass through types
- Add custom converters for attrs tuple conversion to aid typing of attrs tuples with automatic converters.

### CODE REVIEW

File: qualtran/_infra/adjoint_test.py [hunk 1]
```diff
@@ -162,8 +162,8 @@ def test_wire_symbol():
     (reg,) = zero.signature
     adj = Adjoint(zero)  # specifically use the Adjoint wrapper for testing
 
-    ws = zero.wire_symbol(Soquet(None, reg))
-    adj_ws = adj.wire_symbol(Soquet(None, reg.adjoint()))
+    ws = zero.wire_symbol(Soquet(BloqInstance(CNOT(), 1), reg))
+    adj_ws = adj.wire_symbol(Soquet(BloqInstance(CNOT(), 2), reg.adjoint()))
```
Comment (mpharrigan) [inline qualtran/_infra/adjoint_test.py:166]: this is #608 and #728; can you add a todo comment that links to #608 and either revert the change or note that the `BloqInstance`s are made up and ought to be ignored

File: qualtran/_infra/bloq_example.py [hunk 2]
```diff
@@ -36,18 +38,18 @@ class BloqExample:
         generalizer: Passed to `get_bloq_counts_graph` calls for bloq-counts equivalence checking.
     """
 
-    _func: Callable[[], Bloq] = field(repr=False, hash=False)
+    _func: Callable[[], BloqType] = field(repr=False, hash=False)
     name: str
     bloq_cls: Type[Bloq]
-    generalizer: Callable[[Bloq], Optional[Bloq]] = field(
+    generalizer: Callable[[BloqType], Optional[BloqType]] = field(
```
Comment (mpharrigan) [inline qualtran/_infra/bloq_example.py:44]: loving the generics, but this one actually shouldn't be `BloqType`. This function can be anything that takes bloqs to bloqs. You can use `qualtran.resource_counting.GeneralizerT` actually. And I guess this should be `Union[GeneralizerT, Sequence[GeneralizerT]]`

File: qualtran/_infra/bloq_example.py [hunk 3]
```diff
@@ -73,28 +75,30 @@ def _bloq_cls_from_func_annotation(func: Callable[[], Bloq]) -> Type[Bloq]:
...
+) -> Callable[[Callable[[], BloqType]], BloqExample[BloqType]]:
     ...
 
 
 def bloq_example(
-    _func: Callable[[], Bloq] = None, *, generalizer: Callable[[Bloq], Optional[Bloq]] = lambda x: x
-):
+    _func: Callable[[], BloqType] = None,
+    *,
+    generalizer: Callable[[BloqType], Optional[BloqType]] = lambda x: x,
```
Comment (mpharrigan) [inline qualtran/_infra/bloq_example.py:92]: ditto

File: qualtran/_infra/composite_bloq.py [hunk 4]
```diff
@@ -49,11 +50,13 @@
     from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
     from qualtran.simulation.classical_sim import ClassicalValT
 
+# NDArrays must be bound to np.generic
+SoquetType = TypeVar('SoquetType', bound=np.generic)
```
Comment (mpharrigan) [inline qualtran/_infra/composite_bloq.py:54]: can we make this private, i.e. `_SoquetType`? or will it be showing up in more places?

File: qualtran/_infra/composite_bloq.py [hunk 5]
```diff
@@ -62,6 +65,16 @@
 """
 
 
+def _to_tuple(x: Iterable[Connection]) -> Sequence[Connection]:
+    """Needed so mypy can correctly infer types."""
```
Comment (mpharrigan) [inline qualtran/_infra/composite_bloq.py:69]: """mypy-compatible attrs converter for CompositeBloq.connections"""

^ suggestion

File: qualtran/bloqs/arithmetic/addition_test.py [hunk 6]
```diff
@@ -272,7 +272,7 @@ def test_add_mod_n(bitsize, mod, add_val, cvs):
     op = gate.on_registers(**get_named_qubits(gate.signature))
     circuit = cirq.Circuit(op)
     cirq.testing.assert_equivalent_computational_basis_map(basis_map, circuit)
-    circuit += op**-1
+    circuit += cirq.inverse(op)
```
Comment (mpharrigan) [inline qualtran/bloqs/arithmetic/addition_test.py:275]: does this work? I think this is a bigger change than you think

File: qualtran/bloqs/chemistry/sparse/prepare.py [hunk 7]
```diff
@@ -102,12 +102,12 @@ def _add(p_indx: int, q_indx: int, r_indx: int, s_indx: int):
...
-    pqrs_indx = np.array(pqrs_indx)
-    keep_indx = np.where(np.abs(eris_eight) > drop_element_thresh)
-    eris_eight = eris_eight[keep_indx]
-    pqrs_indx = pqrs_indx[keep_indx[0]]
-    return np.concatenate((tpq_indx, pqrs_indx)), np.concatenate((tpq_sparse, eris_eight))
+    eris_eight_np = np.array(eris_eight)
+    pqrs_indx_np = np.array(pqrs_indx)
+    keep_indx = np.where(np.abs(eris_eight_np) > drop_element_thresh)
+    eris_eight_np = eris_eight_np[keep_indx]
+    pqrs_indx_np = pqrs_indx_np[keep_indx[0]]
```
Comment (mpharrigan) [inline qualtran/bloqs/chemistry/sparse/prepare.py:109]: this sort of thing shouldn't be necessary with mypy `allow_redefinition = true` which should already be in the mypy.ini file; which I actually feel quite strongly about :)

File: qualtran/cirq_interop/_cirq_to_bloq.py [hunk 8]
```diff
@@ -56,8 +56,13 @@
 if TYPE_CHECKING:
     from qualtran.drawing import WireSymbol
 
-CirqQuregT = NDArray[cirq.Qid]  # type: ignore[type-var]
-CirqQuregInT = Union[NDArray[cirq.Qid], Sequence[cirq.Qid]]  # type: ignore[type-var]
+
+# numpy subtypes must be np.generic
+# However, this denotes a numpy array of type cirq.Qid
+QidType = TypeVar('QidType', bound=np.generic)
```
Comment (mpharrigan) [inline qualtran/cirq_interop/_cirq_to_bloq.py:62]: can this be private?

File: qualtran/_infra/bloq_example.py (hunk 2 above)
Comment (dstrain115) [reply qualtran/_infra/bloq_example.py:44]: Done.  Let me know if I still have it incorrect.

File: qualtran/_infra/bloq_example.py (hunk 3 above)
Comment (dstrain115) [reply qualtran/_infra/bloq_example.py:92]: Done.

File: qualtran/_infra/composite_bloq.py (hunk 4 above)
Comment (dstrain115) [reply qualtran/_infra/composite_bloq.py:54]: Done.

File: qualtran/_infra/composite_bloq.py (hunk 5 above)
Comment (dstrain115) [reply qualtran/_infra/composite_bloq.py:69]: Done.

**Review by mpharrigan — COMMENTED** [review]
Can you revert the `**-1 -> cirq.inverse()` changes. This will complicate the long-simmering effort to disentangle us from cirq. `__pow__` is a method we have control over and can seamlessly remain supported when we move `GateWithRegisters -> Bloq`

File: qualtran/bloqs/arithmetic/addition_test.py (hunk 6 above)
Comment (dstrain115) [reply qualtran/bloqs/arithmetic/addition_test.py:275]: Why do you say so?  I think it does the same thing in a slightly safer (and more type friendly) way.

`"Returns the inverse `val**-1` of the given value, if defined"
`
https://github.com/quantumlib/Cirq/blob/v1.3.0/cirq-core/cirq/protocols/inverse_protocol.py#L70-L121

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 7 above)
Comment (dstrain115) [reply qualtran/bloqs/chemistry/sparse/prepare.py:109]: Hmmm... it was giving me an error and I was using the mypy.ini which has` "allow_redefinition = true"` in it.  I guess I could investigate further.

File: qualtran/cirq_interop/_cirq_to_bloq.py (hunk 8 above)
Comment (dstrain115) [reply qualtran/cirq_interop/_cirq_to_bloq.py]: Done.

**Comment by dstrain115** [comment]
> Can you revert the `**-1 -> cirq.inverse()` changes. This will complicate the long-simmering effort to disentangle us from cirq. `__pow__` is a method we have control over and can seamlessly remain supported when we move `GateWithRegisters -> Bloq`

cirq.inverse removed

**Comment by dstrain115** [comment]
Ok, I think this is ready for re-review.  Only exception is the variable renaming / redefintion.

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 7 above)
Comment (mpharrigan) [reply qualtran/bloqs/chemistry/sparse/prepare.py]: haha I appreciate it

**Review by mpharrigan — APPROVED** [review]
